src/diffusers/models/hooks.py

```diff
@@ -0,0 +1,255 @@
+
+    def detach_hook(self, module):
+        for hook in self.hooks:
+            module = hook.detach_hook(module)
+        return module
+
+    def reset_state(self, module):
+        for hook in self.hooks:
+            module = hook.reset_state(module)
+        return module
+
+
+class PyramidAttentionBroadcastHook(ModelHook):
+    _stateful_hook = True
+
+    def __init__(
+        self,
+        skip_range: int,
+        timestep_range: Tuple[int, int],
+        timestep_callback: Callable[[], Union[torch.LongTensor, int]],
+    ) -> None:
+        super().__init__()
+
+        self.skip_range = skip_range
+        self.timestep_range = timestep_range
+        self.timestep_callback = timestep_callback
+
+        self.attention_cache = None
+        self._iteration = 0
+
+    def new_forward(self, module: torch.nn.Module, *args, **kwargs) -> Any:
+        args, kwargs = module._diffusers_hook.pre_forward(module, *args, **kwargs)
+
+        current_timestep = self.timestep_callback()
+        is_within_timestep_range = self.timestep_range[0] < current_timestep < self.timestep_range[1]
+        should_compute_attention = self._iteration % self.skip_range == 0
+
+        if not is_within_timestep_range or should_compute_attention:
+            output = module._old_forward(*args, **kwargs)
+        else:
+            output = self.attention_cache
+
+        self._iteration = self._iteration + 1
+
+        return module._diffusers_hook.post_forward(module, output)
+
+    def post_forward(self, module: torch.nn.Module, output: Any) -> Any:
+        self.attention_cache = output
+        return output
+
+    def reset_state(self, module: torch.nn.Module) -> torch.nn.Module:
+        self.attention_cache = None
+        self._iteration = 0
+        return module
+
+
+def add_hook_to_module(module: torch.nn.Module, hook: ModelHook, append: bool = False):
+    r"""
+    Adds a hook to a given module. This will rewrite the `forward` method of the module to include the hook, to remove
+    this behavior and restore the original `forward` method, use `remove_hook_from_module`.
+
+    <Tip warning={true}>
+
+    If the module already contains a hook, this will replace it with the new hook passed by default. To chain two hooks
+    together, pass `append=True`, so it chains the current and new hook into an instance of the `SequentialHook` class.
+
+    </Tip>
+
+    Args:
+        module (`torch.nn.Module`):
+            The module to attach a hook to.
+        hook (`ModelHook`):
+            The hook to attach.
+        append (`bool`, *optional*, defaults to `False`):
+            Whether the hook should be chained with an existing one (if module already contains a hook) or not.
+
+    Returns:
+        `torch.nn.Module`:
+            The same module, with the hook attached (the module is modified in place, so the result can be discarded).
+    """
+    original_hook = hook
+
+    if append and getattr(module, "_diffusers_hook", None) is not None:
```

Comment:
@a-r-r-o-w you mentioned this method takes care of setting `_diffusers_hook`. I would have expected to see something like `setattr("_diffusers_hook")` which I don't. What am I missing?

Reply:
Please see the lines that follow. We create `_diffusers_hook` attribute by assignment if it doesn't exist, or re-use if it does

Reply:
Maybe I am still missing something obvious here but ` old_hook = module._diffusers_hook` does assume we have the `_diffusers_hook` attribute no?

Reply:
Yes, because the check for `getattr(module, "_diffusers_hook", None) is not None` would pass, so we can access it as expected.

For reference, let me try explaining with the code:

```python
    # This branch is only entered if we already added diffusers hook for something else in the past
    # and now also want to sequentially use another hook (that is, combining something like partial offloading
    # and PAB
    if append and getattr(module, "_diffusers_hook", None) is not None:
        old_hook = module._diffusers_hook
        remove_hook_from_module(module)
        hook = SequentialHook(old_hook, hook)
    
    # If we have a previous diffusers hook, we need to make sure to call the previously wrapped forward
    if hasattr(module, "_diffusers_hook") and hasattr(module, "_old_forward"):
        # If we already put some hook on this module, we replace it with the new one.
        old_forward = module._old_forward
    
    # If this is the first diffusers hook that is being added, we will need to make sure to wrap the forward call
    else:
        old_forward = module.forward
        module._old_forward = old_forward

    module = hook.init_hook(module)

    # 1. If an old hook was present but append=False, then we will overwrite the hook
    # 2. If an old hook was present but append=True, then we will overwrite the hook with a SequentialHook combining previous and current hook
    # 3. If an old hook was absent, then regardless of append=True/False, we will create an attribute on the module with the current hook
    module._diffusers_hook = hook
```

LMK if I'm missing something that needs better explanation - it's pretty much 1:1 from accelerate except for added support of `new_forward` methods in the class

Reply:
Ah now it makes sense. Thanks for explaining!